Django samples

waprin opened the pull request.

Commit:
Add Django Samples

Contains two sets of samples - one for the
“Google auth” system and one for the “Django
user auth” system.

nathanielmanistaatgoogle commented on samples/django/README.md:
```diff
@@ -0,0 +1,17 @@
+# Django Samples
+
+These two sample Django apps provide a skeleton for the two main use cases of the
+Django contrib helpers
```
"contrib" isn't an English word but it is a term of art we use in our code. Consider putting it in ticks so that it appears as `contrib`?

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/README.md:
```diff
@@ -0,0 +1,17 @@
+# Django Samples
+
+These two sample Django apps provide a skeleton for the two main use cases of the
+Django contrib helpers
```
Missing period on this line (it looks funny when I read it [rendered](https://github.com/waprin/oauth2client/tree/e13c1f2f6f9d59746a537ae8439bd1e99d49494e/samples/django)).

nathanielmanistaatgoogle commented on samples/django/README.md:
```diff
@@ -0,0 +1,17 @@
+
+This is the simpler use case of the library and assumes you are using Google OAuth as your primary
+authorization and authentication mechanism for your app. Users log in with their Google ID and 
+their OAauth2 credentials are stored inside the session. Please see the core docs for 
```
"the core docs" should be a hyperlink to the core docs.

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/README.md:
```diff
@@ -0,0 +1,17 @@
+ 
+This is the use case where the application is already using the Django authorization system and
+has a Django model with a `django.contrib.auth.models.User` field, and would like to attach
+a Google OAuth2 credentials object to that model. Please see the core docs for usage examples.
```
Smells like copypasta; either correct the sentence or link the words "the core docs"?

waprin replied:
Moved the copy/pasted part to the top.

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
+
+ALLOWED_HOSTS = []
```
Never write a list literal where a tuple literal will do (especially for constants). Does this have to be a list?

waprin replied:
95% of the file is auto-generated by Django. In most cases it does in fact have to be the type it's set to, and even if it doesn't, probably better to stick to conventions.

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+
+ROOT_URLCONF = 'myoauth.urls'
+
+TEMPLATES = [
```
Does this have to be a list?

waprin replied:
see other comment

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+
+GOOGLE_OAUTH2_CLIENT_SECRET = 'YOUR_CLIENT_SECRET'
+
+GOOGLE_OAUTH2_SCOPES = ('email', 'https://www.googleapis.com/auth/calendar',
```
Line-wrap after the opening parenthesis so that all elements of this sequence appear on one line.

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,31 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from django.conf.urls import include, url
```
[Import only packages and modules (mostly modules)](https://google.github.io/styleguide/pyguide.html#Imports).

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,31 @@
+
+from django.conf.urls import include, url
+from django.contrib import admin
+import django.contrib.auth.views
```
[Use `from <package hierarchy> import <module>`](https://google.github.io/styleguide/pyguide.html#Imports).

... unless there's such a thing as a Django style guide that is in conflict with the Google Python Style Guide _and_ it's more important in these examples to demonstrate Django-compliant style?

waprin replied:
Ah was reading down and now I see you mentioning this. If you feel strongly about sticking to `oauth2client` style we can. Following type conventions is obviously more important than following import conventions. I don't think it's super important to demonstrate Django-compliant style, and I don't even necessarily think importing methods is a strict convention although Django does do it in its tutorials. In any case given this is clearly not part of the `oauth2client` package and is a Django sample my inclination is to follow Django conventions but in places where it really can go either way like import conventions, don't feel strongly about it and can change it.

nathanielmanistaatgoogle replied:
Please conform to [Google style](https://google.github.io/styleguide/pyguide.html) for two reasons: (1) this is Google-authored code, and (2) "this code demonstrates use of library <x>, therefore it should be implemented in the style in which library <x> was implemented" strikes me as inappropriate coupling, nonsense, and not something that I find at all persuasive. I feel a little bit negotiable if you want to push back, though?

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,31 @@
+from oauth2client.contrib.django_util.site import urls as oauth2_urls
+
+
+urlpatterns = [
```
Can this be a tuple? If it is intended to be a constant, can it be in all-caps?

waprin replied:
> Django loads that Python module and looks for the variable urlpatterns. This should be a Python list of django.conf.urls.url() instances.

https://docs.djangoproject.com/en/1.10/topics/http/urls/#naming-url-patterns

¯_(ツ)_/¯

nathanielmanistaatgoogle replied:
I am saddened.

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,31 @@
+    url(r'^login', django.contrib.auth.views.login, name="login")
+]
+
+urlpatterns += [url(r'^oauth2/', include(oauth2_urls))]
```
Why add this separately rather than include it in the initial value?

waprin replied:
I'll change this, the reason it's like this is because in the docs it makes the snippet more copy-pastable but not a good reason to keep it like that here.

nathanielmanistaatgoogle commented on samples/django/django_user/polls/views.py:
```diff
@@ -0,0 +1,44 @@
+        return HttpResponse('User email: {}'.format(
+            request.oauth.credentials.id_token['email']))
+    else:
+        return HttpResponse('Here is an OAuth Authorize link:'
```
Avoid implicit string concatenation whenever possible. Line-wrap after the opening parenthesis here?

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/django_user/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Django==0.10.0
+oauth2client==
```
I don't think I understand this line?

waprin replied:
just a mistake, fixed

nathanielmanistaatgoogle commented on samples/django/google_user/myoauth/settings.py:
```diff
@@ -0,0 +1,109 @@
+
+GOOGLE_OAUTH2_CLIENT_SECRET = 'YOUR_CLIENT_SECRET'
+
+GOOGLE_OAUTH2_SCOPES = ('email', 'https://www.googleapis.com/auth/calendar',
```
Line-wrap after opening parenthesis.

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/google_user/myoauth/urls.py:
```diff
@@ -0,0 +1,27 @@
+    url(r'^profile_enabled$', get_profile_optional),
+]
+
+urlpatterns += [url(r'^oauth2/', include(oauth2_urls))]
```
Why is this not included in the initial statement of the value?

waprin replied:
same as above, done

nathanielmanistaatgoogle commented on samples/django/google_user/polls/templates/registration/login.html:
```diff
@@ -0,0 +1,47 @@
+    <input type="submit" value="login" />
+    <input type="hidden" name="next" value="{{ next }}" />
+</form>
+
```
Are these extra newlines at the end of the file significant in any way? Can they be dropped?

waprin replied:
no, done

nathanielmanistaatgoogle commented on samples/django/google_user/polls/views.py:
```diff
@@ -0,0 +1,45 @@
+        return HttpResponse('User email: {}'.format(
+            request.oauth.credentials.id_token['email']))
+    else:
+        return HttpResponse('Here is an OAuth Authorize link:'
```
Avoid implicit string concatenation whenever you can.

waprin replied:
done

waprin commented:
@nathanielmanistaatgoogle responded to all comments but need to take another pass myself but have to run at this minute, will finish up later tonight. Will also stop importing functions across the samples though as discussed debatable whether it makes more sense to stick with Django conventions over oauth2client conventions.

waprin commented:
@nathanielmanistaatgoogle ok ready for another pass, also cleaned up some unnecessary files. Following the style guide for imports does make sense to me, I think it's just the types in `settings.py` we should leave alone. 

Whenever review is LGTM, besides needing to squash the review commits, I'd like to do a final pass on an actual server for both these samples before merge.

cloud-dpe-jenkins commented:
Starting a new build on Jenkins!

cloud-dpe-jenkins commented:
Build on Jenkins Started!

cloud-dpe-jenkins commented:
Build finished. No test results found.

waprin commented:
Sorry experimenting with Jenkins job to report system tests in background.

wasabigeek commented on samples/django/django_user/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Django==0.10.0
```
Should this be 1.10.0?

waprin replied:
yes thanks. manually wrote it out since my virtualenv has some unrelated stuff, hopefully would have caught this in final walkthrough. Thank you for taking look.

wasabigeek replied:
Cool. I haven't had the chance to actually try the code yet, will update when I do.

Thanks a lot for updating the samples!

wasabigeek commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+    'django.middleware.common.CommonMiddleware',
+    'django.middleware.csrf.CsrfViewMiddleware',
+    'django.contrib.auth.middleware.AuthenticationMiddleware',
+    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
```
hi @waprin, quick question - I assume this is taken from the tutorial as when I did `djangoadmin startproject` this did not turn up, although it still [works](https://code.djangoproject.com/ticket/23786). Is it necessary?

waprin replied:
Thanks, done

wasabigeek commented on samples/django/google_user/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Django==1.10.0
+oauth2client==3.0.0
```
I think this will not work out-of-the-box till https://github.com/google/oauth2client/issues/593 is released

wasabigeek commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,29 @@
+    urls.url(r'^profile_enabled$', views.get_profile_optional),
+    urls.url(r'^admin/', urls.include(admin.site.urls)),
+    urls.url(r'^login', django.contrib.auth.views.login, name="login"),
+    urls.url(r'^oauth2/', urls.include(django_util.site.urls))
```
`AttributeError: 'module' object has no attribute 'site'`

waprin replied:
fixed

wasabigeek commented on samples/django/django_user/myoauth/urls.py:
```diff
@@ -0,0 +1,29 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from django.conf.urls import urls
```
Should we follow more closely what `djangoadmin startproject` generates i.e. `from django.conf.urls import url, include` instead of importing the whole module?

waprin replied:
per Nathaniel's comments, Google Python Style is to only import packages/modules. I think we are following Google Python Style over Django style except where it would cause too much trouble.

wasabigeek commented on samples/django/django_user/polls/templates/registration/login.html:
```diff
@@ -0,0 +1,45 @@
+{% endif %}
+{% endif %}
+
+<form method="post" action="{% url 'django.contrib.auth.views.login' %}">
```
```
NoReverseMatch at /login
Reverse for 'django.contrib.auth.views.login' with arguments '()' and keyword arguments '{}' not found. 0 pattern(s) tried: []
```

{% url 'login' %} works fine

waprin replied:
thanks, done

wasabigeek commented on samples/django/django_user/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Django==1.10.0
+oauth2client==3.0.0
```
This version doesn't work with 1.10 as it still doesn't include #593 and https://github.com/google/oauth2client/issues/594

waprin replied:
Yeah there will be a new release soon with all those fixes

wasabigeek commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+GOOGLE_OAUTH2_CLIENT_SECRET = 'YOUR_CLIENT_SECRET'
+
+GOOGLE_OAUTH2_SCOPES = (
+    'email', 'https://www.googleapis.com/auth/calendar', 'profile')
```
Scopes requested don't match https://github.com/google/oauth2client/pull/636/files#diff-de2fc5853437bc87b3335ca6d764cf40R27

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/django_user/manage.py:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
```
Does this file need a license header?

waprin replied:
I was unsure since its 100% generated but apparently yes

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+    'django.contrib.staticfiles',
+    'polls',
+    'oauth2client.contrib.django_util'
+
```
Spurious blank line?

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+    'django.contrib.messages',
+    'django.contrib.staticfiles',
+    'polls',
+    'oauth2client.contrib.django_util'
```
For multi-line sequence literals: comma after the last element? I have a slight preference for "yes"; do you care either way?

waprin replied:
Getting those comments in other reviews and seems harmless so yes I will start doing it, done.

nathanielmanistaatgoogle commented on samples/django/django_user/myoauth/settings.py:
```diff
@@ -0,0 +1,117 @@
+GOOGLE_OAUTH2_STORAGE_MODEL = {
+    'model': 'polls.models.CredentialsModel',
+    'user_property': 'user_id',
+    'credentials_property': 'credential'
```
Comma after last key-value pair in this dictionary literal value?

waprin replied:
done

nathanielmanistaatgoogle commented on samples/django/google_user/myoauth/settings.py:
```diff
@@ -0,0 +1,109 @@
+    'django.contrib.staticfiles',
+    'polls',
+    'oauth2client.contrib.django_util'
+
```
Spurious blank line?

nathanielmanistaatgoogle commented:
Completed another review pass; I don't think I noticed anything beyond cosmetics.

wasabigeek commented on samples/django/django_user/polls/views.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@decorators.oauth_required
+def get_profile_required(request):
```
TypeError at /oauth2/oauth2authorize/
' ' is not JSON serializable (Python 3.4.4)

wasabigeek replied:
Sorry ignore, realised it's related to https://github.com/google/oauth2client/pull/636/files#r77437699

wasabigeek commented on samples/django/django_user/polls/views.py:
```diff
@@ -0,0 +1,41 @@
+            request.oauth.credentials.id_token['email']))
+    else:
+        return HttpResponse(
+            'Here is an OAuth Authorize link:<a href="%s">Authorize</a>'
```
`{0}` instead of `%s`?

waprin replied:
Done

wasabigeek commented on samples/django/django_user/polls/models.py:
```diff
@@ -0,0 +1,34 @@
+class Choice(models.Model):
+    question = models.ForeignKey(Question, on_delete=models.CASCADE)
+    choice_text = models.CharField(max_length=200)
+    votes = models.IntegerField(default=0)
```
Both models are unnecessary?

waprin replied:
Hmm I can kill them, the motivation to have them was to the Django polls tutorial as the context for the samples, but agreed less code is better so I killed them.

wasabigeek commented:
@waprin did you have the chance to take a look at the comments?

waprin commented:
@wasabigeek was on vacation from work and programming in general, will look back around on this tonight!

waprin commented:
@wasabigeek @nathanielmanistaatgoogle thanks for review comments, I addressed all the issues and tested them again, ready for another pass at review pass. Prob should wait until a release before we merge though since the samples working do depend on some of the bugfixes.

nathanielmanistaatgoogle commented:
I'm satisfied. Who else will be reviewing this?

waprin commented:
I think we had an oauth2client maintainer and someone who tried the samples (thanks again @wasabigeek ) so we should be good to merge. I will also at some point take a look at the google-api-python-client and see if it makes sense to redirect here which should get us more eyes. But since the samples will have known bug fixes until the next release, maybe @jonparrott wants to merge whenever next release is cut (after we update the requirements.txt)?

jonparrott commented:
@waprin the next release is cut for what? this library? What changes are needed?

waprin commented:
@jonparrott #623 #651 #614 . Otherwise the samples are unfortunately probably too buggy to even have in the repo.

jonparrott commented:
I'm going to merge this and cut a new release.

jonparrott commented:
Actually, I can't release without bumping a major version which we wanted to reserve for the transport refactors.

@waprin where are these samples referenced?

waprin commented:
@jonparrott they are only referenced here for now

jonparrott commented:
@waprin then I am okay with the situation for now. Let's hold off on any external references until we can release 4.0.0.
